src/shutdown.rs

```diff
@@ -0,0 +1,253 @@
+    }
+
+    /// Takes ownership of all internal state for the given source from another ShutdownCoordinator.
+    pub fn takeover_source(&mut self, name: &str, other: &mut Self) {
+        self.shutdown_begun_triggers.insert(
+            name.to_string(),
+            other.shutdown_begun_triggers.remove(name).unwrap(),
+        );
+
+        self.shutdown_force_triggers.insert(
+            name.to_string(),
+            other.shutdown_force_triggers.remove(name).unwrap(),
+        );
+
+        self.shutdown_complete_tripwires.insert(
+            name.to_string(),
+            other.shutdown_complete_tripwires.remove(name).unwrap(),
+        );
+    }
+
+    pub fn shutdown_source_begin(&mut self, name: &str) {
+        self.shutdown_begun_triggers.remove(name).unwrap().cancel();
+    }
+
+    /// Waits for the source to shut down until the deadline.  If the source does not
+    /// notify the shutdown_complete_tripwire for this source before the dealine, then signals
+    /// the shutdown_force_trigger for this source to force it to shut down.  Returns whether
+    /// or not the source shutdown gracefully.
+    // TODO: The timing and reporting logic is very similar to the logic in
+    // `RunningTopology::stop()`. Once `RunningTopology::stop()` has been updated to utilize the
+    // ShutdownCoordinator, see if some of this logic can be de-duped.
+    pub fn shutdown_source_end<'a>(
+        &mut self,
+        rt: &mut runtime::Runtime,
+        name: String,
+        deadline: Instant,
+    ) -> bool {
+        let name2 = name.clone();
+        let name3 = name.clone();
+        let shutdown_complete_tripwire = self.shutdown_complete_tripwires.remove(&name).unwrap();
+        let shutdown_force_trigger = self.shutdown_force_triggers.remove(&name).unwrap();
+
+        let success = shutdown_complete_tripwire.map(move |_| {
+            info!("Source \"{}\" shut down successfully", name);
+        });
+        let timeout = timer::Delay::new(deadline)
+            .map(move |_| {
+                error!(
+                    "Source '{}' failed to shutdown before deadline. Forcing shutdown.",
+                    name2,
+                );
+            })
+            .map_err(|err| panic!("Timer error: {:?}", err));
+        let reporter = timer::Interval::new_interval(Duration::from_secs(5))
+            .inspect(move |_| {
+                let time_remaining = if deadline > Instant::now() {
+                    format!(
+                        "{} seconds remaining",
+                        (deadline - Instant::now()).as_secs()
+                    )
+                } else {
+                    "overdue".to_string()
+                };
+
+                info!(
+                    "Still waiting on source \"{}\" to shut down. {}",
+                    name3, time_remaining,
+                );
+            })
+            .filter(|_| false) // Run indefinitely without emitting items
```

Comment:
This doesn't make sense to me, shouldn't this basically yield no items ever?

Reply:
Without the filter, this is a stream of `Instant`s.